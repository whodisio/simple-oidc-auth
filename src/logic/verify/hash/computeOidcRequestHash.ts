import { sha256 } from 'cross-sha256';

/**
 * computes a one-way hash of the inputs of the request
 *
 * usecase
 * - this hash can be later compared against to verify that
 *   - the responder had access to the original request inputs
 *   - the backend knows exactly which inputs the request was made with
 */
export const computeOidcRequestHash = ({
  oidcRequestUuid,
  userSessionUuid,
}: {
  /**
   * the primary key identifier of the request, persisted by your backend
   */
  oidcRequestUuid: string;

  /**
   * the primary key identifier of the current user's session, if any
   */
  userSessionUuid: string | null;
}): string =>
  [
    'orh',
    new sha256()
      .update(
        JSON.stringify({
          purpose: 'oidcRequestHash',
          oidcRequestUuid,
          userSessionUuid,
        }),
      )
      .digest('hex'),
  ].join('_');
