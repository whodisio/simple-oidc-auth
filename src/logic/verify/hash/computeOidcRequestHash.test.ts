import { uuid } from '../../../deps';
import { computeOidcRequestHash } from './computeOidcRequestHash';

describe('computeOidcRequestHash', () => {
  it('should be deterministic', () => {
    const inputs = {
      oidcRequestUuid: uuid(),
      userSessionUuid: null,
    };
    expect(computeOidcRequestHash(inputs)).toEqual(
      computeOidcRequestHash(inputs),
    );
  });
  it('should look nice', () => {
    const inputs = {
      oidcRequestUuid: '0fb30413-49ab-483d-8b8e-19e3d3e96eec',
      userSessionUuid: null,
    };
    expect(computeOidcRequestHash(inputs)).toMatch(/^orh_/); // starts with orh
    expect(computeOidcRequestHash(inputs)).toMatchSnapshot();
  });
});
