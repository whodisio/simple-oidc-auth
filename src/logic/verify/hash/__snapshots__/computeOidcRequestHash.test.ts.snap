// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeOidcRequestHash should look nice 1`] = `"orh_df207b150a2e7ad900aabee8d4aa64e07a6ba19be8cd972a51a824fde1707cac"`;
